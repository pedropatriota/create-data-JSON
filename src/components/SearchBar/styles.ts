import styled from "styled-components";

export const SearchContainer = styled.div<{ expand: string }>`
  width: 100%;
  max-width: 34rem;
  height: ${({ expand }) => (expand === "true" ? "30rem" : "4rem")};
  display: flex;
  flex-direction: column;
  margin-top: 3rem;
  background-color: #fff;
  overflow: hidden;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.2);
  border-radius: 0.5rem;
  transition: height 0.5s ease-in-out;
`;

export const InputContainer = styled.div`
  width: 100%;
  min-height: 4rem;
  display: flex;
  align-items: center;
  position: relative;
  padding: 0 1rem;
`;

export const Input = styled.input`
  width: 100%;
  height: 100%;
  outline: none;
  border: none;
  background-color: transparent;
  font-size: 1.2rem;
  font-weight: 500;

  &:focus {
    outline: none;
    border-bottom: 1px solid #ccc;
  }
  &::placeholder {
    color: #ccc;
    font-size: 1rem;
  }
`;

export const CloseIcon = styled.span<{ inputvalue: string }>`
  color: #ccc;
  opacity: ${({ inputvalue }) => (inputvalue ? 1 : 0)};
  transition: opacity 0.5s ease-in-out;
  cursor: pointer;
`;

export const DropdownContainer = styled.ul`
  overflow-y: auto;
  padding-left: 1rem;
`;
