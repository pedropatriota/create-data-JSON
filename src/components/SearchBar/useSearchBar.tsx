import { useState, useEffect, useCallback } from "react";
import { searchData } from "../../service/fetchRequest";
import type { ICharacterProps } from "../../service/contracts";
import { useClickOutside, useDebounce } from "../../hooks";

const useSearchBar = () => {
  const [expand, setExpand] = useState(false);
  const [inputValue, setInputValue] = useState("");
  const [result, setResult] = useState<ICharacterProps[] | undefined>([]);

  const ref = useClickOutside(() => setExpand(false));
  const debouncedValue = useDebounce(inputValue, 200);

  useEffect(() => {
    if (debouncedValue) {
      searchData(debouncedValue)
        .then((data) => setResult(data))
        .catch((err) => console.log(err));
    } else {
      setResult([]);
    }
  }, [debouncedValue]);

  const expandSearchBox = useCallback(() => {
    setExpand(true);
  }, []);

  const collapseSearchBox = useCallback(() => {
    setExpand(false);
  }, []);

  const handleChangeInputValue = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(event.target.value);
    },
    []
  );

  const clearInputValue = useCallback(() => {
    setInputValue("");
    collapseSearchBox();
    setResult([]);
  }, []);

  return {
    expand,
    inputValue,
    result,
    ref,
    expandSearchBox,
    handleChangeInputValue,
    clearInputValue,
  };
};

export default useSearchBar;
