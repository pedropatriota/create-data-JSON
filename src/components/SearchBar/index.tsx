import { Character } from "..";
import useSearchBar from "./useSearchBar";
import * as Styled from "./styles";

const SearchBar = () => {
  const {
    expand,
    inputValue,
    result,
    ref,
    expandSearchBox,
    handleChangeInputValue,
    clearInputValue,
  } = useSearchBar();

  return (
    <Styled.SearchContainer
      expand={expand.toString()}
      ref={ref}
      data-testid="search-container"
    >
      <Styled.InputContainer>
        <Styled.Input
          onFocus={expandSearchBox}
          onChange={handleChangeInputValue}
          value={inputValue}
          placeholder="Make your search..."
        />
        {inputValue && (
          <Styled.CloseIcon onClick={clearInputValue}>
            &#10006;
          </Styled.CloseIcon>
        )}
      </Styled.InputContainer>
      <Styled.DropdownContainer>
        {result?.map(({ name, id, image }) => (
          <Character
            key={id}
            name={name}
            image={image}
            inputValue={inputValue}
          />
        ))}
      </Styled.DropdownContainer>
    </Styled.SearchContainer>
  );
};

export default SearchBar;
