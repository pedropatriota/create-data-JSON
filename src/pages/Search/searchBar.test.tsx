import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type { IRepoProps } from "../../service/contracts";
import SearchBar from "./searchBar";

interface SearchBar {
  expand: boolean;
  inputValue: string;
  result: IRepoProps[];
  loading: boolean;
}

const setup = ({ expand, inputValue, result, loading }: SearchBar) => {
  const expandSearchBox = jest.fn();
  const handleChangeInputValue = jest.fn();
  const clearInputValue = jest.fn();

  const props = {
    expand,
    inputValue,
    result,
    expandSearchBox,
    handleChangeInputValue,
    clearInputValue,
    loading,
  };
  const utils = render(<SearchBar {...props} />);
  return {
    ...utils,
    expandSearchBox,
    handleChangeInputValue,
    clearInputValue,
    props,
  };
};

afterEach(() => cleanup);

describe("SearchBar", () => {
  const repo = {
    id: "1",
    name: "Example",
    url: "https://exemplo.com/",
    owner: {
      avatarUrl: "https://github.com/example",
      login: "owner",
    },
    description: "descriptionExample",
  };
  it("should render the initial state correctly", async () => {
    setup({ expand: false, inputValue: "", result: [], loading: false });

    const input = (await screen.findByRole("textbox")) as HTMLInputElement;
    const searchContainer = screen.getByTestId("search-container");
    const closeIcon = screen.queryByText(/✖/i);

    expect(input).toBeInTheDocument();
    expect(input).not.toHaveFocus();
    expect(input.value).toBe("");
    expect(searchContainer).toHaveStyle("height: 4rem");
    expect(closeIcon).not.toBeInTheDocument();
  });

  it("should render correctly after focus", async () => {
    const { container } = setup({
      expand: true,
      inputValue: "",
      result: [],
      loading: false,
    });

    const input = (await screen.findByRole("textbox")) as HTMLInputElement;
    const searchContainer = screen.getByTestId("search-container");

    await userEvent.click(input);
    fireEvent.focus(input);

    expect(searchContainer).toHaveStyle("height: 25rem");
    expect(input).toHaveFocus();
  });

  it("should be able type in the input and fetch the API", async () => {
    setup({
      expand: true,
      inputValue: "Example",
      result: [{ ...repo }],
      loading: false,
    });

    const input = (await screen.findByRole("textbox")) as HTMLInputElement;
    const searchContainer = screen.getByTestId("search-container");

    await userEvent.click(input);
    fireEvent.focus(input);

    expect(searchContainer).toHaveStyle("height: 25rem");
    expect(input).toHaveFocus();

    await userEvent.type(input, "Example");
    expect(input.value).toBe("Example");

    const name = await screen.findAllByText("Example");

    const closeIcon = screen.queryByText(/✖/i);

    expect(closeIcon).toBeInTheDocument();
    expect(name.length).toBe(1);
  });

  it("should collapse the dropdown", async () => {
    const {
      rerender,
      expandSearchBox,
      handleChangeInputValue,
      clearInputValue,
    } = setup({
      expand: true,
      inputValue: "Example",
      result: [{ ...repo }],
      loading: false,
    });

    const input = (await screen.findByRole("textbox")) as HTMLInputElement;
    const searchContainer = screen.getByTestId("search-container");

    await userEvent.click(input);
    fireEvent.focus(input);
    expect(searchContainer).toHaveStyle("height: 25rem");
    await userEvent.type(input, "Example");

    const name = await screen.findAllByText("Example");
    const closeIcon = await screen.findByText(/✖/i);

    await userEvent.click(closeIcon);

    rerender(
      <SearchBar
        expand={false}
        inputValue=""
        result={[]}
        expandSearchBox={expandSearchBox}
        handleChangeInputValue={handleChangeInputValue}
        clearInputValue={clearInputValue}
        loading={false}
      />
    );

    expect(input.value).toBe("");
    expect(searchContainer).toHaveStyle("height: 4rem");
  });

  it("should match snapshot", () => {
    const { container } = setup({
      expand: true,
      inputValue: "Example",
      result: [],
      loading: false,
    });

    expect(container).toMatchSnapshot();
  });
});
