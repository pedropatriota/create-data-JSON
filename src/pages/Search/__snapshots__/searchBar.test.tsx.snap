// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<div>
  <img
    alt="logo GitHub"
    class="sc-imWYAI ewKPJB"
    src="test-file-stub"
  />
  <div
    class="sc-fqkvVR hlFEFn"
    data-testid="search-container"
  >
    <div
      class="sc-dcJsrY MyhEQ"
    >
      <input
        class="sc-iGgWBj bAOAVV"
        placeholder="Make your search..."
        value="Example"
      />
      <span
        class="sc-gsFSXq dBTFVK"
      >
        âœ–
      </span>
    </div>
    <ul
      class="sc-kAyceB cXmXts"
    />
  </div>
</div>
`;
